### ISSUE TITLE
Cast from int? to int is incorrectly marked as redundant

### ISSUE THREAD

**Issue by vsfeedback** [dotnet/roslyn#22804]
see attachement

_This issue has been moved from https://developercommunity.visualstudio.com/content/problem/118692/cast-redundant-but-its-not.html
VSTS ticketId: 499179_
_These are the original issue comments:_

David Poeschl [MSFT] on 9/25/2017, 00:41 PM (27 days ago): <p>Thank you for your feedback! Unfortunately, I can't find the attachment you mention. Can you upload it again or just paste the code into a comment? Thanks!</p>

_These are the original issue solutions:_

Clemens Schneider solved on 9/25/2017, 01:38 PM (27 days ago), 0 votes: 
```C#
using System.Text;
namespace ConsoleApp2 
{ 
    class Program 
    { 
        static void Main(string[] args) 
        { 
            StringBuilder numbers = new StringBuilder(); 
            int?[] position = new int?[2];
            numbers[(int)position[1]] = 'x'; 
        } 
    } 
} 
```

**Comment by dpoeschl** [comment]
Formatted code example:
```C#
using System.Text;
namespace ConsoleApp2
{
    class Program
    {
        static void Main(string[] args)
        {
            StringBuilder numbers = new StringBuilder();
            int?[] position = new int?[2]; numbers[($$int)position[1]] = 'x';
        }
    }
}
```
Incorrect "redundant cast" is at the $$.

**Comment by jcouv** [comment]
@dpoeschl You didn't know that github auto-formats. (j/k)

**Comment by dpoeschl** [comment]
Did you update the OP? I just added the "C#" to the code block. I only really made it a separate post so the @vsfeedback folks could see how things auto-formatted, but they can figure it out based on how it looked on developercommunity.

**Comment by jcouv** [comment]
Hehe, I did. Not stalking you or anything, I promise ;-)

**Comment by pdelvo** [comment]
After a lot of debugging I think I tracked down the root cause:

- The analyzer uses a SpeculativeSyntaxModel to check wether or not changing `numbers[(int)position[1]]` to `numbers[position[1]]` would change the semantics.

- While trying to bind `numbers[position[1]]` it correctly notices the overload resolution error.
- This is marked by setting OriginalIndexersOpt to the right (not default) value.
- On the way up at some point BoundIndexerAccess.Update is [called](https://github.com/dotnet/roslyn/blob/4e109f8d7bab5072fd0488bc5ccd4dec1fda13e8/src/Compilers/CSharp/Portable/Binder/Binder.ValueChecks.cs#L190) which just drops the value because the property is not part of the auto generated code. From this point forward the information that overload resolution failed is (except for the error list) lost.

I think the easiest solution would be to add OriginalIndexersOpt to the automatically generated part of the code. I prototyped this and it seems to fix the issue: https://github.com/dotnet/roslyn/commit/91a7494bc245a91461fe0d0cbf55632e811beee3 .

Is this the right approach? If it is I can open a PR for this that includes tests.

**Comment by pdelvo** [comment]
I was going to fix this but it looks like this PR: #23028 already fixed the issue.

**Comment by CyrusNajmabadi** [comment]
This doesn't repro anymore.  Will add validating test.
